docs/architecture/adr-003-domain-decomposition.md

```diff
@@ -0,0 +1,165 @@
+The goal of this ADR is to provide clarity around the domain objects,
+their interfaces as well as dependencies between options in order to
+guide the implementation. The success criteria for the decomposition is
+how well it can be tested. It's expected that any decomposition will
+lend itself to tight unit tests allowing more collaborators make change
+in the code base with confidence.
+
+## Decision
+The decomposition should be motivated by what we want to tests and what
+do we need to mock out to exercise the core logic.
+
+* Connection handshake
+    * With different chain states
+* Channel Setup
+    * With different chain states
+* Datagram Construction
+    * Local lighht client state
+    * With different chain states
+    * With different light client states
+* Datagram to Transaction
+    * batching
+    * Signing
+* Datagram Submission
+    * With different chain states
+
+* Handlers (datagrams, chain state) -> events
+    * batch of datagrams and chain states
+        * Specifically, the key value store
+    * Produce events
+
+* Operational concerns
+    * CLI, embedded
+
+## Objects
+The main domain object will be implemented as a trait. This will allow
+dependencies between objects to be mocked out during testing.
+
+### Chain
+The Chain object is a local representation of a foreign chain state.
+The API of a Chain provides reliable access to chain state Wether
+cashed, constructed or queried.
+
+```rust 
+struct ChainState {
+    // That represents the data that is stored on chain
+    // It's expected that the IBC handlers are producing such a state
+    // It's expected the the Chain object and queries are effectively
+    // caching such a state
+    ...
+}
+
+```
+
+The actual chain trait will be implemented by a simple synchronous
+version at first and offer a mocked alternative for testing.
+
+```rust
+trait Chain {
+    // Query and potentially cache consensus_state etc
+    fn query_*(&mut self, height: Height, ...) -> Result<..., Error> {
+        ...
+    }
+
+    // Or maybe we should just provide setters and getters and allow
+
+    // Chains have have embeded local light clients to verify headers
+    fn verify(&mut self, header: Header, ...) -> Result<..., Error> {
+    }
+
+    // 
+    fn submit(&mut self, datagrams: vec<Dataram>) -> Result<..., ...> {
+        // We can check if the packet is indeed worth submitting here
+        ...
+        // We can ensure that the remote client is up to date before submitting
+        self.update_remote_client();
+        ...
+    }
+
+    // Produce a subscription of verified events
+    fn subscribe(&mut self, ...) -> Subscription {
+        // Create a subscription to events in this chains runtime whith
+        // access to the chain such that it can verify packets as they
+        // come in
+        return Subscription::new(self.clone(), ...);
+    }
+}
+```
+
+### Connection
+
+```rust
+trait Connection {
```

Comment:
Should these really be traits? IBC objects like Connection, Channel, Packet are all concrete types that don't vary across chains. Clients are more of an interface and do vary depending on the chain, so we end up with concrete types like the Tendermint Client, Grandpa Client, Local Client, etc.

Reply:
Agreed, if those don't vary across chain then they should definitely be plain structs, and we leave it to the traits to abstract over operations, etc. rather than data.